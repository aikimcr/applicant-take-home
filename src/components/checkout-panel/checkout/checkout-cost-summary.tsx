import React from 'react';
import { PrizeoutOfferValueOptions } from '../../../slices/offers-slice';

import './checkout.less';

interface CheckoutCostSummaryProps {
    giftcard: PrizeoutOfferValueOptions;
}

export const CheckoutCostSummary: React.FC<CheckoutCostSummaryProps> = ({ giftcard }): React.ReactElement => {
    const redemptionAmount = (giftcard.cost_in_cents / 100).toFixed(2);
    const finalValue = (giftcard.value_in_cents / 100).toFixed(2);

    function bonusLine() {
        let displayPercent = '';
        let bonus = null;

        if (!giftcard.display_bonus && !giftcard.display_monetary_bonus) {
            return '';
        }
        if (!giftcard.display_bonus) {
            bonus = giftcard.display_monetary_bonus / 100;
        } else {
            const bonusPercent = giftcard.display_bonus / 100;
            displayPercent = `(${giftcard.display_bonus}%)`;
            bonus = Math.round(giftcard.cost_in_cents * bonusPercent) / 100;
        }

        return (
            <>
                <label className="bonus-line">{`Prizeout Bonus${displayPercent}`}</label>
                <div className="value bonus-line">${bonus.toFixed(2)}</div>
            </>
        );
    }

    return (
        <section className="checkout-cost-summary">
            <div className="grid">
                <label>Redemption Amount</label>
                <div className="value">${redemptionAmount}</div>
                {bonusLine()}
                <label>You Get</label>
                <div className="value">{finalValue}</div>
            </div>
        </section>
    );
};

export default CheckoutCostSummary;
