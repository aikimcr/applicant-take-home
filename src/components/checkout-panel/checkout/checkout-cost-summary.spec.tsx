import { render, screen } from '../../../testing/test-utils';
import React from 'react';
import '@testing-library/jest-dom';

import { CheckoutCostSummary } from './checkout-cost-summary';
import { PrizeoutOfferValueOptions } from '../../../slices/offers-slice';
import { click } from '@testing-library/user-event/dist/click';

describe('Test CheckoutCostSummary', () => {
    test('Component matches snapshot', () => {
        const testCard: PrizeoutOfferValueOptions = {
            checkout_value_id: 'xyzzy',
            cost_in_cents: 1212,
            display_bonus: 33,
            display_monetary_bonus: null,
            value_in_cents: 1612,
        };

        const { asFragment } = render(<CheckoutCostSummary giftcard={testCard} />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('show display_bonus (percentage) if it is defined', () => {
        const testCard: PrizeoutOfferValueOptions = {
            checkout_value_id: 'xyzzy',
            cost_in_cents: 1212,
            display_bonus: 33,
            display_monetary_bonus: null,
            value_in_cents: 1612,
        };

        render(<CheckoutCostSummary giftcard={testCard} />);

        expect(screen.getByText('Prizeout Bonus(33%)')).toBeVisible();
        expect(screen.getByText('$4.00')).toBeVisible();
    });

    test('show display_monetary bonus if it is defined', () => {
        const testCard: PrizeoutOfferValueOptions = {
            checkout_value_id: 'xyzzy',
            cost_in_cents: 1212,
            display_bonus: null,
            display_monetary_bonus: 200,
            value_in_cents: 1412,
        };

        render(<CheckoutCostSummary giftcard={testCard} />);

        expect(screen.getByText('Prizeout Bonus')).toBeVisible();
        expect(screen.getByText('$2.00')).toBeVisible();
    });

    test('show display_monetary bonus if display_bonus is zero', () => {
        const testCard: PrizeoutOfferValueOptions = {
            checkout_value_id: 'xyzzy',
            cost_in_cents: 1212,
            display_bonus: 0,
            display_monetary_bonus: 200,
            value_in_cents: 1412,
        };

        render(<CheckoutCostSummary giftcard={testCard} />);

        expect(screen.getByText('Prizeout Bonus')).toBeVisible();
        expect(screen.getByText('$2.00')).toBeVisible();
    });

    test('hide the bonus line if neither value is defined', () => {
        const testCard: PrizeoutOfferValueOptions = {
            checkout_value_id: 'xyzzy',
            cost_in_cents: 1212,
            display_bonus: null,
            display_monetary_bonus: null,
            value_in_cents: 1212,
        };

        render(<CheckoutCostSummary giftcard={testCard} />);

        expect(screen.queryByText(/Prizeout Bonus/)).not.toBeInTheDocument();
    });

    test('hide the bonus line if both values are zero', () => {
        const testCard: PrizeoutOfferValueOptions = {
            checkout_value_id: 'xyzzy',
            cost_in_cents: 1212,
            display_bonus: 0,
            display_monetary_bonus: 0,
            value_in_cents: 1212,
        };

        render(<CheckoutCostSummary giftcard={testCard} />);

        expect(screen.queryByText(/Prizeout Bonus/)).not.toBeInTheDocument();
        expect(screen.queryByText('$0.00')).not.toBeInTheDocument();
    });
});
