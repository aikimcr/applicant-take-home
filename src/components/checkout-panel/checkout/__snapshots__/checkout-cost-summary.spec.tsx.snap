// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CheckoutCostSummary Component matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="checkout-cost-summary"
  >
    <div
      class="grid"
    >
      <label>
        Redemption Amount
      </label>
      <div
        class="value"
      >
        $12.12
      </div>
      <label
        class="bonus-line"
      >
        Prizeout Bonus(33%)
      </label>
      <div
        class="value bonus-line"
      >
        $4.00
      </div>
      <label>
        You Get
      </label>
      <div
        class="value"
      >
        16.12
      </div>
    </div>
  </section>
</DocumentFragment>
`;
